obsidianirc.c

```diff
@@ -0,0 +1,640 @@
+        {
+                "The module is installed. Now all you need to do is add a loadmodule line:";
+                "loadmodule \"third/obsidian-register\";";
+                "The module needs no other configuration.";
+                "Once you're good to go, you can finally type in your shell: ./unrealircd rehash";
+        }
+}
+*** <<<MODULE MANAGER END>>>
+*/
+
+/* One include for all */
+#include "obsidian.h"
+
+long CAP_ACCOUNTREGISTRATION = 0L;
+
+ModuleHeader MOD_HEADER
+= 
+{
+    "third/obsidianirc", /* Name of module */
+    "1.0", /* Version */
+    "ObsidianIRC", /* Short description of module */
+    "ObsidianIRC Team", /* Author */
+    "unrealircd-6", /* Version of UnrealIRCd */
+};
+
+/**
+ * MOD_INIT - Module initialization routine.
+ */
+MOD_INIT()
+{
+    ModDataInfo mreq;
+
+    memset(&mreq, 0, sizeof(mreq));
+    mreq.name = "sasl_auth_type";
+    mreq.free = sat_free;
+    mreq.serialize = sat_serialize;
+    mreq.unserialize = sat_unserialize;
+    mreq.type = MODDATATYPE_CLIENT;
+    if (!(sasl_md = ModDataAdd(modinfo->handle, mreq)))
+    {
+        config_error("Could not add ModData for sasl_auth_type. Please contact developer.");
+        return MOD_FAILED;
+    }
+
+    if (open_database(OBSIDIAN_DB) != SQLITE_OK)
+    {
+        config_error("Could not open database. Please contact ObsidianIRC Support.");
+        return MOD_FAILED;
+    }
+
+    ClientCapabilityInfo accreg_cap; 
+    memset(&accreg_cap, 0, sizeof(accreg_cap));
+    accreg_cap.name = REGCAP_NAME;
+    accreg_cap.visible = accreg_capability_visible;
+    accreg_cap.parameter = accreg_capability_parameter;
+    if (!ClientCapabilityAdd(modinfo->handle, &accreg_cap, &CAP_ACCOUNTREGISTRATION))
+    {
+        config_error("Could not add CAP for draft/account-registration. Please contact ObsidianIRC Support.");
+        return MOD_FAILED;
+    }
+
+    HookAddConstString(modinfo->handle, HOOKTYPE_SASL_MECHS, 0, saslmechs);
+    HookAdd(modinfo->handle, HOOKTYPE_SASL_AUTHENTICATE, 0, authenticate_attempt);
+    CommandAdd(modinfo->handle, CMD_REGISTER, register_account, 3, CMD_USER|CMD_UNREGISTERED);
+    CommandAdd(modinfo->handle, CMD_LISTACC, list_accounts, 3, CMD_OPER);
+    return MOD_SUCCESS;
+}
+
+MOD_LOAD()
+{
+    if (open_database(OBSIDIAN_DB) != SQLITE_OK)
+    {
+        config_error("Could not open database. Please contact ObsidianIRC Support.");
+        return MOD_FAILED;
+    }
+    safe_strdup(iConf.sasl_server, me.name);
+    moddata_client_set(&me, "saslmechlist", "PLAIN,EXTERNAL");
+    return MOD_SUCCESS;
+}
+
+/**
+ * MOD_UNLOAD - Called when the module is unloaded.
+ */
+MOD_UNLOAD()
+{
+    close_database();
+    safe_free(iConf.sasl_server);
+    iConf.sasl_server = NULL;
+    return MOD_SUCCESS;
+}
+
+/**
+ * register_account - Handles the REGISTER command from users.
+ * Syntax: REGISTER <name> <email> <password>
+ */
+CMD_FUNC(register_account)
+{
+    if (!db && open_database(OBSIDIAN_DB) != SQLITE_OK)
```

Comment:
Do we need to open database every time, wont sqlite fail if multiple users do it at the same time or is this. I am not sure about this